File: src/contracts/clinicalEvent.ts
import type { InputModality, ISODate } from './common'

export type ClinicalEntityType = 'Disease' | 'Symptom' | 'Medication' | 'Allergy' | 'Procedure' | 'LabFinding' | 'LaboratoryFinding' | 'Dosage' | 'Route'
export type ClinicalAssertion = 'affirmed' | 'negated' | 'possible'
export type ClinicalStatus = 'active' | 'historical' | 'resolved' | 'unknown'
export type TemporalContextType = 'current' | 'historical' | 'future' | 'specific_date'
export type ClinicalValidationStatus = 'valid' | 'invalid_schema' | 'invalid_entity' | 'invalid_relationship' | 'incomplete_provenance'

export interface ClinicalRelationship { relation_type: string; target_event_local_id: string }
export interface AmbiguousAbbreviationResolution { was_ambiguous: boolean; resolved_value: string | null; resolution_confidence: number }
export interface MedicationAttributes { dosage: string | null; strength: string | null; frequency: string | null; route: string | null }
export interface LabAttributes { test_name: string | null; value: string | null; unit: string | null }

export interface ClinicalEvent {
  event_local_id: string
  original_text: string
  processed_text: string
  normalized_concept: string
  snomed_ct_id: string | null
  entity_type: ClinicalEntityType
  clinical_domain: string
  relationships: ClinicalRelationship[]
  assertion: ClinicalAssertion
  clinical_status: ClinicalStatus
  temporal_context: TemporalContextType
  temporal_date: ISODate | null
  bioclinicalbert_confidence: number
  gemini_contextualization_confidence: number
  ambiguous_abbreviation_resolved: AmbiguousAbbreviationResolution
  source_document_id: string
  source_text_span: { start: number; end: number }
  input_modality: InputModality
  source_language: string
  translation_confidence: number | null
  medication_attributes: MedicationAttributes
  lab_attributes: LabAttributes
  validation_status: ClinicalValidationStatus
}

export interface Step2Response {
  clinical_events: ClinicalEvent[]
  patient_id: string
  encounter_id: string
  source_document_id: string
  processed_at: ISODate
}

File: contracts/schemas/clinical_event.py
from datetime import date, datetime, timezone
from typing import Any, Self
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .enums import ClinicalEventValidationStatus, InputModality


class SourceTextSpan(BaseModel):
    model_config = ConfigDict(extra="forbid")

    start: int = Field(ge=0)
    end: int = Field(ge=0)

    @model_validator(mode="after")
    def validate_order(self) -> Self:
        if self.end < self.start:
            raise ValueError("Text span end must not precede its start.")
        return self


class ClinicalEvent(BaseModel):
    """Exact Step 2 -> Step 3 event contract from the backend handoff."""

    model_config = ConfigDict(extra="forbid")

    event_local_id: UUID
    original_text: str = Field(min_length=1)
    processed_text: str = Field(min_length=1)
    normalized_concept: str = Field(min_length=1)
    snomed_ct_id: str | None = None
    entity_type: str = Field(min_length=1)
    clinical_domain: str = Field(min_length=1)
    relationships: list[dict[str, Any]] = Field(default_factory=list)
    assertion: str = Field(min_length=1)
    clinical_status: str = Field(min_length=1)
    temporal_context: str = Field(min_length=1)
    temporal_date: date | None = None
    bioclinicalbert_confidence: float = Field(ge=0.0, le=1.0)
    gemini_contextualization_confidence: float = Field(ge=0.0, le=1.0)
    source_document_id: UUID
    source_text_span: SourceTextSpan
    input_modality: InputModality
    source_language: str = Field(min_length=1, max_length=20)
    translation_confidence: float = Field(ge=0.0, le=1.0)
    validation_status: ClinicalEventValidationStatus


class ClinicalEventBatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    clinical_events: list[ClinicalEvent]
    patient_id: UUID
    encounter_id: UUID
    source_document_id: UUID
    processed_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
    )
